### PR TITLE
CCIP Read ISM

### PR DESCRIPTION
**Pull request by AlexBHarley**
### Description

CCIP Read ISM implementation,

- [x] Contracts
- [x] Relayer
- [x] Example HTTP backend

### Drive-by changes

_Are there any minor or drive-by changes also included?_

### Related issues

- Fixes #[2211]

### Backward compatibility

_Are these changes backward compatible?_

Yes, relayers will skip messages they can't handle.

_Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?_

Yes, relayers won't be able to submit messages that have a CCIP Read ISM.

### Testing

_What kind of testing have these changes undergone?_

- [x] Manual
- [x] Unit Tests

### General approach

- [x] Get a basic [CCIP Read implementation working](https://github.com/AlexBHarley/permissionless-chainlink-feeds/tree/main/contracts)
- [x] Build a [mock service that serves CCIP Read compatible data](https://github.com/AlexBHarley/permissionless-chainlink-feeds/tree/main/apps/api)
- [x] Update the Relayer to support CCIP Read ISMs

### CODE REVIEW

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (yorhodes) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:132]: ideally the abstract CCIP ism doesnt know about signatures or digests
that being said I understand you wanted to just get something working

**Comment by AlexBHarley** [comment]
After chatting with @nambrot last week I've taken a slightly different approach to the original one, interested to hear what you guys think. An implementation of a CcipReadIsm can be found [here](https://github.com/AlexBHarley/permissionless-chainlink-feeds/).

I have a couple open questions I'm not sure about,

- Is there any value in the `AbstractCcipReadIsm` living in the monorepo? I feel it's OK for consumers to implement this logic themselves, but on the other hand it's nice to have the `OffchainLookup` error well-typed for use in the relayer
- Do we actually want to call this `CcipRead`? While it's quite close to matching the spec, it doesn't exactly and that could confuse client side implementations. e.g., relayer queries for information on the destination chain, but the callback function (Mailbox.process) is not only a different contract but a different chain. Simply calling it `OffchainRead` could be better
- If we do keep with CCIP Read, using the custom error approach, I'm unsure how to better parse these. [This section of the code](https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2398/files#diff-5d0308f2ded7da5abaff992e4e8eee32845b360c1a9f522ff2f9bdd4c8a14b59R54) is what I'm referring to. Right now I do a Regex match to pull the error out which feels messy

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (nambrot) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:46]: Does this need to be specified here? IMO there is no need to prescribe that this data is stored in storage right?

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (nambrot) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:16]: Might be useful to comment here the slightly off-spec way of using this event?

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (nambrot) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:81]: Couldn't you get full compliance by calling specifying a function called `process(metadata, message)` which would just call the mailbox (which would call the ISM)?

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (nambrot) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:70]: This should be _message right?

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (nambrot) [reply solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:81]: In that case you probably should make the mailbox part of this interface? Though IMO its probably fine for the contracts that are meant for this use case to not be full spec-compliant

File: solidity/contracts/test/TestCcipReadIsm.sol
Comment (nambrot) [inline solidity/contracts/test/TestCcipReadIsm.sol:12]: Is this still intentional here?

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (nambrot) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:82]: IMO, consumers should explicitly have to specify this (and not rely on the abstract contract passing something from storage (which very well may be uninitialized). If you want to, you can  make this into an internal function that a consumer has to override

File: rust/agents/relayer/src/msg/metadata/ccip_read.rs
Comment (nambrot) [inline rust/agents/relayer/src/msg/metadata/ccip_read.rs:53]: Probably should have send this to you earlier, but have you seen https://github.com/ensdomains/ethers-ccip-read. I wonder if we could reuse logic from there (looks possible?)

I think @yorhodes and I had chatted that the requirement in the spec to require the `data` field in the JSON response is not great, but services like CCTP actually are already compliant https://developers.circle.com/stablecoin/reference/getattestation

Comment (nambrot) [reply rust/agents/relayer/src/msg/metadata/ccip_read.rs:53]: (Might have to fork it and have it return the calldata of the final call)

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (nambrot) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:1]: I assume you were planning on it, but a bit more of a README to show how folks should implement this contract would be nice

File: solidity/contracts/mock/MockMailbox.sol
Comment (nambrot) [inline solidity/contracts/mock/MockMailbox.sol:16]: Was this neccessary?

**Comment by nambrot** [comment]
> * Is there any value in the `AbstractCcipReadIsm` living in the monorepo? I feel it's OK for consumers to implement this logic themselves, but on the other hand it's nice to have the `OffchainLookup` error well-typed for use in the relayer

For sure, IMO its critical to have it in the monorepo as its the reference for the agent implementation as well. Like i commented, it should also include instructions for the spec of this ISM, i.e. how do I override it, what interface does my offchain service have to comply with, etc. Like another comment of mine, the consumer should be forced to implement the extra data of the revert message though.

> * Do we actually want to call this `CcipRead`? While it's quite close to matching the spec, it doesn't exactly and that could confuse client side implementations. e.g., relayer queries for information on the destination chain, but the callback function (Mailbox.process) is not only a different contract but a different chain. Simply calling it `OffchainRead` could be better

Hmm, i'm not sure that I'm following. The relayer would query the ISM on the destination chain for the `OffchainRead` revert message, to then call the Mailbox.process on that very same chain no? (You could even have the ISM call Mailbox.process if you wanted to, to be fully compliant) 

> * If we do keep with CCIP Read, using the custom error approach, I'm unsure how to better parse these. [This section of the code](https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2398/files#diff-5d0308f2ded7da5abaff992e4e8eee32845b360c1a9f522ff2f9bdd4c8a14b59R54) is what I'm referring to. Right now I do a Regex match to pull the error out which feels messy

You should be able to look at this https://github.com/ensdomains/ethers-ccip-read/blob/main/src/middleware.rs#L243C1-L244C1

**Comment by nambrot** [comment]
Also putting a review here for the ChainlinkISM since there is no PR to look at:
- It would be nice to see a diff of the original OffchainAggregator and yours. I have terrible memory, but remind me why the Aggregator even has to be modified at all? I thought we were good with making storage writes in the verify function?
- Are you using `HyperlaneConnectionClient` just to have a storage variable for the mailbox? In which case it might be preferable to not have that dependency.
- Why is `verify` permission to `onlyMailbox`?
- Are etherscan keys required? IMO etherscan lets you do 1 request every 5 seconds which should maybe be fine?
- It would be nice to link to a deployment somewhere (for both the service, and the feed on an EVM chain, and show it replicating via etherscan view calls)

File: solidity/contracts/mock/MockMailbox.sol
Comment (AlexBHarley) [reply solidity/contracts/mock/MockMailbox.sol]: Yes, the Chainlink messages are too big

Comment (AlexBHarley) [reply solidity/contracts/mock/MockMailbox.sol:16]: True... Yes I will revert this

**Comment by nambrot** [comment]
@yorhodes called out that an alternate path for this implementation is to put the CCIP read middleware at the provider level and then have estimate gas to `mailbox.process` just automatically populate the metadata. However, that is probably more involved as there is no out of the box middleware that can be used for this purpose (yet). @AlexBHarley 's approach is sufficient as it enables ISM development without relayer changes, and is technically possible to be spec compliant, and we can always introduce the alternate way in the future

File: solidity/contracts/Mailbox.sol
Comment (nambrot) [inline solidity/contracts/Mailbox.sol:33]: Undo this

File: solidity/contracts/interfaces/IInterchainSecurityModule.sol
Comment (nambrot) [inline solidity/contracts/interfaces/IInterchainSecurityModule.sol:12]: FYI https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2420/files#diff-7d0323c3aa18eb73dd096474183d65c5d2495f0ea62685db5e6f7117c96e63fcR12

Comment (AlexBHarley) [reply solidity/contracts/interfaces/IInterchainSecurityModule.sol:12]: Oh yep, which do we think will be merged first? This or that PR

File: rust/agents/relayer/src/msg/metadata/ccip_read.rs
Comment (AlexBHarley) [reply rust/agents/relayer/src/msg/metadata/ccip_read.rs:55]: For posterity, it seemed a little complicated to use [ethers-ccip-read](https://github.com/ensdomains/ethers-ccip-read) in the relayer. The types used internally don't line up terribly well, and for a simple GET/POST call I also didn't think it's worth bringing in an extra dep

File: solidity/contracts/interfaces/IInterchainSecurityModule.sol
Comment (nambrot) [reply solidity/contracts/interfaces/IInterchainSecurityModule.sol:12]: There is no rush on either of them, asked @yorhodes to take a peek at this one before we merge

File: rust/agents/relayer/src/msg/metadata/ccip_read.rs
Comment (yorhodes) [reply rust/agents/relayer/src/msg/metadata/ccip_read.rs:55]: I think it's a bit more complicated than implemented here
eg supporting recursive offchain lookups
happy to punt for now but this feels like a good place to minimize surface area

File: solidity/contracts/interfaces/isms/ICcipReadIsm.sol
Comment (yorhodes) [inline solidity/contracts/interfaces/isms/ICcipReadIsm.sol:12]: why have a return value at all?

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (yorhodes) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:53]: why?

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (yorhodes) [inline solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:24]: IIUC we wouldn't need to do this "pointer indirection" because the `Mailbox.process` would bubble up the revert reason from the `ism.verify` call

File: solidity/package.json
Comment (yorhodes) [inline solidity/package.json:6]: ?

**Review by yorhodes — APPROVED** [review]
big unlock for new ISMs 🚀

File: solidity/contracts/interfaces/isms/ICcipReadIsm.sol
Comment (AlexBHarley) [reply solidity/contracts/interfaces/isms/ICcipReadIsm.sol:12]: Good point, no idea

File: solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol
Comment (AlexBHarley) [reply solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:24]: Let me be a little pedantic, spec wise, for a second. There's two ways for CCIP Read contracts to work,

1. Call some function, ie. `getOffchainVerifyInfo` which describes how to get and submit the off chain info to a callback
2. Call the function you actually want to call, ie. `Mailbox.process`, which reverts in the absence of off chain data but describes how to get the data. Allowing the caller to then call the function again.

I opted for option 1 because the Mailbox is not CCIP Read aware, so we break a core invariant in CCIP Read that states `If the sender field does not match the address of the contract that was called, return an error to the caller and stop.`. The Mailbox would need to catch the ISM error and revert with its own `OffchainLookup` error in turn.

The Relayer doesn't respect the callback function anyway so it's kind of irrelevant. But IMO option 1 at least has the option of being totally spec compliant if we want

Comment (yorhodes) [reply solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol:24]: that checks out
I do still prefer option 2 in theory but requiring the Mailbox to be CCIP read aware is quite a departure from the expected scope

**Pull request merged**
